File: app/src/main/java/com/cmgapps/android/personalappwidget/AppPackageUpdateReceiver.java
/*
 * Copyright (c) 2022. Christian Grach <christian.grach@cmgapps.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.cmgapps.android.personalappwidget;

import android.appwidget.AppWidgetManager;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;


public class AppPackageUpdateReceiver extends BroadcastReceiver {
    private static final String TAG = "AppPackageUpdateReceiver";
    public static final String ACTION_APP_ADDED = "com.cmgapps.android.personalappwidget.action.APP_ADDED";

    @Override
    public void onReceive(Context context, Intent intent) {
        Log.d(TAG, "onReceive " + intent);

        if (intent.getAction().equals(ACTION_APP_ADDED)) {
            // TODO handle app selection
            Log.d(TAG, intent.getData().getSchemeSpecificPart());
            return;
        }

        Uri data = intent.getData();

        for (String pkg : PersonalAppLauncherWidgetProvider.PACKAGES) {
            if (pkg.equals(data.getSchemeSpecificPart())) {
                AppWidgetManager manager = AppWidgetManager.getInstance(context);

                PersonalAppLauncherWidgetProvider.updateAppWidget(context, manager,
                        manager.getAppWidgetIds(new ComponentName(context, PersonalAppLauncherWidgetProvider.class)));
                break;
            }
        }
    }

}

File: app/src/main/java/com/cmgapps/android/personalappwidget/widget/AppPackageUpdateReceiver.kt
/*
 * Copyright (c) 2022. Christian Grach <christian.grach@cmgapps.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package com.cmgapps.android.personalappwidget.widget

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log
import com.cmgapps.LogTag
import com.cmgapps.android.personalappwidget.BuildConfig
import dagger.hilt.android.AndroidEntryPoint

@LogTag("AppUpdateReceiver")
@AndroidEntryPoint
class AppPackageUpdateReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        if (BuildConfig.DEBUG) {
            Log.d(LOG_TAG, "onReceive $intent")
        }

        when (intent.action) {
            Intent.ACTION_PACKAGE_ADDED, Intent.ACTION_PACKAGE_REMOVED ->
                FavoriteAppWidgetProvider.refreshWidgets(context)
        }
    }
}
